## Упражнения по библиотеке Numpy

In [170]:
import numpy as np

In [171]:
rng = np.random.default_rng(seed=42)

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [172]:

arr1 = 15 * rng.random(50) 
print(arr1)

previous_zero_mask = (arr1 >= 3) & (arr1 <= 8)
arr1[previous_zero_mask] *= -1
print(arr1)


[11.60934073  6.5831766  12.8789688  10.46052044  1.41266022 14.63433527
 11.41709553 11.79096458  1.92170449  6.75578907  5.56197036 13.90147483
  9.6579768  12.3414242   6.65121298  3.40858083  8.31877181  0.95725884
 12.41446758  9.47496599 11.3713161   5.31788952 14.56047037 13.39681682
 11.67575246  2.91958062  7.00081506  0.65705649  2.31434238 10.2457343
 11.17143234 14.51264599  4.88738037  5.55689559  7.04333717  2.84207039
  1.94882258  7.13557389  3.40364024 10.04720992  6.55727878 12.49017294
 10.50397653  4.68549962 12.48389702 12.07146536  5.81217569  4.32492156
 10.23743256  2.09628725]
[11.60934073 -6.5831766  12.8789688  10.46052044  1.41266022 14.63433527
 11.41709553 11.79096458  1.92170449 -6.75578907 -5.56197036 13.90147483
  9.6579768  12.3414242  -6.65121298 -3.40858083  8.31877181  0.95725884
 12.41446758  9.47496599 11.3713161  -5.31788952 14.56047037 13.39681682
 11.67575246  2.91958062 -7.00081506  0.65705649  2.31434238 10.2457343
 11.17143234 14.51264599 -4

**2.** Заменить максимальный элемент случайного массива на 0

In [173]:
# your code
arr2 = rng.random(10) * 5
print(arr2)

max_idx = np.argmax(arr2)

print(max_idx, arr2[max_idx])

arr2[max_idx] = 0    

print(arr2)

[0.99954101 0.03681135 3.93462189 3.32425428 3.52582689 3.90364516
 2.29457888 2.84370598 0.69898499 0.57265037]
2 3.934621887510692
[0.99954101 0.03681135 0.         3.32425428 3.52582689 3.90364516
 2.29457888 2.84370598 0.69898499 0.57265037]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [174]:
# your code
arr3 = rng.integers(0, 10, size=(3, 2))
print(arr3)

grid = np.meshgrid(*arr3, indexing='ij')
res = np.stack(grid, axis=-1).reshape(-1, len(arr3))

print(res)


[[4 6]
 [6 4]
 [8 5]]
[[4 6 8]
 [4 6 5]
 [4 4 8]
 [4 4 5]
 [6 6 8]
 [6 6 5]
 [6 4 8]
 [6 4 5]]


In [175]:
print('Hello', flush=True)
print('World', flush=True)

Hello
World


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [176]:
# your code

a = rng.integers(0, 5, size=(8, 3))
b = rng.integers(0, 5, size=(2, 2))

print(a, '\n\n', b, '\n')

c = (a[..., np.newaxis, np.newaxis] == b)
rows = (c.sum(axis=(1,2,3)) >= b.shape[1]).nonzero()[0]

print(rows)




[[0 3 2]
 [3 2 2]
 [0 2 3]
 [1 3 0]
 [1 2 4]
 [1 1 2]
 [4 4 0]
 [1 4 0]] 

 [[4 1]
 [4 1]] 

[3 4 5 6 7]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [177]:
# your code
m = rng.integers(1, 3, size=(10, 3))
print(m)

even_rows = np.logical_and.reduce(m[:,1:] == m[:,:-1], axis=1)

indices = np.where(~even_rows)[0]

print(indices)


[[1 2 1]
 [2 2 2]
 [2 2 1]
 [1 1 2]
 [1 1 1]
 [1 1 1]
 [2 2 2]
 [1 2 1]
 [2 2 2]
 [1 1 2]]
[0 2 3 7 9]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [178]:
# your code
arr6 = rng.integers(1, 3, size=(10, 3))

print(arr6)

_, unique_indices = np.unique(arr6, axis=0, return_index=True)

sorted_unique_indices = np.sort(unique_indices)

arr6 = arr6[np.sort(sorted_unique_indices)]

print(sorted_unique_indices)
print(arr6)

[[1 1 1]
 [1 1 1]
 [2 2 2]
 [1 2 1]
 [1 1 1]
 [2 1 2]
 [1 2 1]
 [1 2 2]
 [1 2 1]
 [2 2 1]]
[0 2 3 5 7 9]
[[1 1 1]
 [2 2 2]
 [1 2 1]
 [2 1 2]
 [1 2 2]
 [2 2 1]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [179]:
# your code
lst1 = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]

diag_mult = lst1[0][0]

for i in range(1, min(len(lst1), len(lst1[0]))):
    if lst1[i][i] != 0:
        diag_mult *= lst1[i][i]

print(diag_mult)

3


In [180]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

d = X.diagonal()
res = d[d != 0].prod()

print(res)

3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [181]:
# your code

x = [1, 2, 2, 4, ]
y = [4, 2, 1, 2]

x.sort()
y.sort()   


same = (x == y)

print(same)

True


In [182]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

x.sort()
y.sort()   

same = np.array_equal(x, y)

print(same)

True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [183]:
# your code
    
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]


max = -int(1e9)
prev_zero = False
for i in range(len(x)):
    if prev_zero:
        if x[i] > max:
            max = x[i]
            
    if x[i] == 0:
        prev_zero = True
    else:
        prev_zero = False
        
print(max)
        
        



5


In [184]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

previous_zero_mask = (x[:-1] == 0)

after_zeros = x[1:][previous_zero_mask]
print(np.max(after_zeros))

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [185]:
# your code
x = [2, 2, 2, 3, 3, 3, 5]

nums = []
counts = []

curr_num = x[0]
curr_count = 0

for i in range(len(x)):
    curr_count += 1
    if i + 1 == len(x) or x[i] != x[i+1]:
        nums.append(x[i])
        counts.append(curr_count)
        curr_count = 0

print(nums, counts)   

[2, 3, 5] [3, 3, 1]


In [186]:
x = np.array([2, 2, 2, 3, 3, 3, 5])


change_indices = np.where(np.diff(x) != 0)[0]

nums = x[np.append(0, change_indices + 1)]

full_indices = np.concatenate([[-1], change_indices, [len(x) - 1]])

counts = np.diff(full_indices)

print(nums, counts)

[2 3 5] [3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [187]:
import numpy as np
from scipy.spatial.distance import cdist
import time

rng = np.random.default_rng()
x = rng.random(size=(500, 20))
y = rng.random(size=(700, 20))

def pairwise_dist_py(X_list, Y_list):
    out = []
    for x_row in X_list:
        row = []
        for y_row in Y_list:
            s = 0.0
            for xi, yi in zip(x_row, y_row):
                d = xi - yi
                s += d * d
            row.append(s ** 0.5)
        out.append(row)
    return out

def pairwise_dist_np(X_arr, Y_arr):
    return np.sqrt(((X_arr[:, None, :] - Y_arr[None, :, :]) ** 2).sum(axis=2))

def pairwise_dist_fast_np(X, Y):
    x2 = np.sum(X**2, axis=1)[:, np.newaxis]
    y2 = np.sum(Y**2, axis=1)[np.newaxis, :]
    xy = X @ Y.T
    return np.sqrt(np.maximum(x2 - 2*xy + y2, 0))

t0 = time.perf_counter()
res_py = pairwise_dist_py(x.tolist(), y.tolist())
t_py = time.perf_counter() - t0

t0 = time.perf_counter()
res_np = pairwise_dist_np(x, y)
t_np = time.perf_counter() - t0

t0 = time.perf_counter()
res_fast_np = pairwise_dist_fast_np(x, y)
t_fast_np = time.perf_counter() - t0

t0 = time.perf_counter()
res_cdist = cdist(x, y, metric="euclidean")
t_cdist = time.perf_counter() - t0

print(f"Equal results: {np.allclose(res_py, res_np, res_fast_np, res_cdist)}")

print(f"Python loops:      {t_py:.4f} s")
print(f"Numpy vectorized:  {t_np:.4f} s")
print(f"Numpy Matrix Mult: {t_fast_np:.4f} s")
print(f"Scipy cdist:       {t_cdist:.4f} s")

Equal results: True
Python loops:      0.7374 s
Numpy vectorized:  0.0539 s
Numpy Matrix Mult: 0.0148 s
Scipy cdist:       0.0075 s


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [188]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [189]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [190]:
# your code
average_calories = calorie_stats.mean()
print(f"Average calories: {average_calories}")

Average calories: 106.88311688311688


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [191]:
# your code
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [192]:
# your code
median_calories = np.median(calorie_stats)
print(f"Median calories: {median_calories}")

Median calories: 110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [197]:
# your code
nth_percentile = 0

for i in range(101):
    current_value = np.percentile(calorie_stats, i)
    print(f"{i}th percentile: {current_value}")
    if current_value > 60:
        nth_percentile = i
        break
print(f"{nth_percentile}th percentile is the first one above 60 calories.")

0th percentile: 50.0
1th percentile: 50.0
2th percentile: 50.0
3th percentile: 55.599999999999994
4th percentile: 70.0
4th percentile is the first one above 60 calories.


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [194]:
# your code
more_calories = calorie_stats[calorie_stats > 60].size / calorie_stats.size * 100
print(f"Percentage of cereals with more than 60 calories: {more_calories:.2f}%")


Percentage of cereals with more than 60 calories: 96.10%


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [195]:
# your code
calorie_std = np.std(calorie_stats)
print(f"Standard deviation of calories: {calorie_std:.2f}")

Standard deviation of calories: 19.36


In [199]:
calorie_moda = np.bincount(calorie_stats.astype(int)).argmax()
print(f"Mode of calories: {calorie_moda}")

Mode of calories: 110


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

По результатам анализа видно, что у конкурентов средняя калорийность составляет 106.9 ккал, медиана мода — 110.0 ккал, то есть рапсределение правоскошенное. Уже с 4-го процентиля значения превышают калорийность CrunchieMunchies - 60 ккал. Доля хлопьев с калорийностью выше 60 ккал достигает 96.1%, а стандартное отклонение 19.4 указывает на заметный разброс, но в целом типичный продукт конкурентов значительно калорийнее. Эти данные можно использовать в маркетинге MyCrunch акцентируя внимание на более лёгкой калорийности, чем у 96 процентов, в том числе с добавлением графика рапсределения и на большом различии калорийности CrunchieMunchies и среднего значения калорийности, например, указанием процентного соотношения.